ada_feeding_perception/ada_feeding_perception/face_detection.py

```diff
@@ -0,0 +1,314 @@
+                        description="The name of the face detection model checkpoint to use",
+                        read_only=True,
+                    ),
+                ),
+                (
+                    "landmark_model_base_url",
+                    None,
+                    ParameterDescriptor(
+                        name="landmark_model_base_url",
+                        type=ParameterType.PARAMETER_STRING,
+                        description="The URL to download the model checkpoint from if it is not already downloaded",
+                        read_only=True,
+                    ),
+                ),
+                (
+                    "model_dir",
+                    None,
+                    ParameterDescriptor(
+                        name="model_dir",
+                        type=ParameterType.PARAMETER_STRING,
+                        description="The location of the directory where the model checkpoint is / should be stored",
+                        read_only=True,
+                    ),
+                ),
+            ],
+        )
+
+    def toggle_face_detection_callback(self, request, response):
+        """
+        Callback function for the ToggleFaceDetection service. Safely toggles
+        the face detection on or off depending on the request.
+        """
+        self.get_logger().info(
+            "Incoming service request. turn_on: %s" % (request.turn_on)
+        )
+        if request.turn_on:
+            # Turn on face detection
+            self.is_on_lock.acquire()
+            self.is_on = True
+            self.is_on_lock.release()
+            response.face_detection_is_on = True
+        else:
+            self.is_on_lock.acquire()
+            self.is_on = False
+            self.is_on_lock.release()
+            response.face_detection_is_on = False
+        return response
+
+    def depth_callback(self, msg):
+        """
+        Callback function for depth images. If a face has been detected in the
+        most recent rgb image, this function publishes the 3d location of the
+        mouth on the /face_detection topic in the camera_depth_optical_frame
+        """
+        self.is_on_lock.acquire()
+        is_on = self.is_on
+        self.is_on_lock.release()
+
+        # Check if face detection is on and a face has been detected
+        if is_on and self.is_face_detected:
+            # Retrieve the 2d location of the mouth center
+            u = int(self.img_mouth_center[0])
+            v = int(self.img_mouth_center[1])
+
+            # Retrieve the depth value for this 2d coordinate
+            width = msg.width
+            depth = msg.data[(u % width) + (v * width)]
+
+            # Create target 3d point
+            mouth_location = PointStamped()
+            mouth_location.header = msg.header
+            mouth_location.point.x = float(u)
+            mouth_location.point.y = float(v)
+            mouth_location.point.z = float(depth)
```

Comment:
The depth image coordinates are in `mm`. However, the norm in ROS is for `PoseStamped` points to be in `m`. Can you convert them before putting them into the message?